### ISSUE TITLE
Serializing std::vector<int> vs std::vector<int> const different sizes

### ISSUE THREAD

**Issue by bjadamson** [eyalz800/zpp_bits#164]
`OS: windows`
`Compiler:
Microsoft Visual Studio Community 2022 (64-bit) - Current
Version 17.10.2`

Hi, is this behavior expected?

I've been reading through the section on serializing [standard library types](https://github.com/eyalz800/zpp_bits?tab=readme-ov-file#standard-library-types-serialization) but I can't find anything anything about whether the vector being const affecting the sizes. What I understand is the size should always be serialized using 4 bytes (by default).

```C++
#include <external/zpp_full.hpp>
#include <cassert>

using U8Buffer = std::vector<std::uint8_t>;
using U8Writer = zpp::bits::out<U8Buffer>;

int main(int, char**)
{
  U8Buffer b0;
  U8Buffer b1;
  {
    U8Writer writer{b0};
    std::vector<int> const data;
    assert(std::errc{} == writer(data));
  }
  {
    U8Writer writer{b1};
    std::vector<int> data;
    assert(std::errc{} == writer(data));
  }
  assert(b0 == b1); // FAIL, std::size(b0)==0, and std::size(b1)==4
  return 0;
}
```

**Comment by bjadamson** [comment]
Here's a link using gcc:
https://godbolt.org/z/h9es5KbYf

**Comment by bjadamson** [comment]
To anyone else who runs into this for now I'm just casting away the `const` from my reference that I'm serializing.

**Comment by eyalz800** [comment]
It's most likely an MSVC bug, if I change this code:
![image](https://github.com/eyalz800/zpp_bits/assets/29372486/59dab2b9-4648-4fb8-b6de-ae4332f004df)

To this:

![image](https://github.com/eyalz800/zpp_bits/assets/29372486/4ebe41ae-54a0-4450-8eda-1fffef3a6ed6)

It resolves - this means that msvc finds the wrong "container" variable despite the nearest variable is the one declared in the requires clause.

**Comment by bjadamson** [comment]
Hi @eyalz800 ,

Thanks for looking into this. Unfortunately the issue seems to happen in gcc/clang as well:

https://github.com/eyalz800/zpp_bits/issues/164#issuecomment-2168793777

I hope this helps. If it helps I noticed the issue also seems to occur with std::variant I'm using.

**Comment by eyalz800** [comment]
@bjadamson but in GCC and clang both are size 4 no?

**Comment by eyalz800** [comment]
But in GCC and clang is says 4 for both in your links:
![image](https://github.com/eyalz800/zpp_bits/assets/29372486/3dc54a86-38c9-497a-9f75-4cf3ef7d5079)

**Comment by bjadamson** [comment]
Wow your right I'm so sorry. My brain was swapping the b0 with the value for some reason.

**Comment by eyalz800** [comment]
@bjadamson at least you seem to have found some bug in MSVC, I didn't submit a patch because my entire code is written with name hiding from requires expressions so fixing just this piece is not gonna cut it - want to submit an MSVC bug and link it here?

**Comment by bjadamson** [comment]
Sure -- if you have suggestions on going through the process I'd love to hear them. This will be my first bug report with those folks.

**Comment by eyalz800** [comment]
Always a good day to learn something new

**Comment by SenseOnline** [comment]
Hi, I couldn't find an issue opened with Microsoft regarding this and went ahead with submitting a ticket. 
https://developercommunity.visualstudio.com/t/C-requires-expression-name-shadowing-i/10726925

**Comment by SenseOnline** [comment]
Updating that this issue has been resolved in MSVC release version 17.12.

**Comment by eyalz800** [comment]
Thank you for the update
